apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/service/AccessKeyServiceWithCache.java

```diff
@@ -0,0 +1,206 @@
+
+    initialize();
+  }
+
+  private void initialize() {
+    scheduledExecutorService = new ScheduledThreadPoolExecutor(1,
+        ApolloThreadFactory.create("AccessKeyServiceWithCache", true));
+    lastTimeScanned = new Date(0L);
+
+    ListMultimap<String, AccessKey> multimap = ListMultimapBuilder.hashKeys(128)
+        .arrayListValues().build();
+    accessKeyCache = Multimaps.synchronizedListMultimap(multimap);
+    accessKeyIdCache = Maps.newConcurrentMap();
+  }
+
+  public List<String> getAvailableSecrets(String appId) {
+    List<AccessKey> accessKeys = accessKeyCache.get(appId);
+    if (CollectionUtils.isEmpty(accessKeys)) {
+      return Collections.emptyList();
+    }
+
+    return accessKeys.stream()
+        .filter(AccessKey::isEnabled)
+        .map(AccessKey::getSecret)
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public void afterPropertiesSet() throws Exception {
+    populateDataBaseInterval();
+    scanNewAndUpdatedAccessKeys(); //block the startup process until load finished
+
+    scheduledExecutorService.scheduleWithFixedDelay(this::scanNewAndUpdatedAccessKeys,
+        scanInterval, scanInterval, scanIntervalTimeUnit);
+
+    scheduledExecutorService.scheduleAtFixedRate(this::rebuildAccessKeyCache,
+        rebuildInterval, rebuildInterval, rebuildIntervalTimeUnit);
+  }
+
+  private void scanNewAndUpdatedAccessKeys() {
+    Transaction transaction = Tracer.newTransaction("Apollo.AccessKeyServiceWithCache",
+        "scanNewAndUpdatedAccessKeys");
+    try {
+      loadNewAndUpdatedAccessKeys();
+      transaction.setStatus(Transaction.SUCCESS);
+    } catch (Throwable ex) {
+      transaction.setStatus(ex);
+      logger.error("Load new/updated app access keys failed", ex);
+    } finally {
+      transaction.complete();
+    }
+  }
+
+  private void rebuildAccessKeyCache() {
+    Transaction transaction = Tracer.newTransaction("Apollo.AccessKeyServiceWithCache",
+        "rebuildCache");
+    try {
+      deleteAccessKeyCache();
+      transaction.setStatus(Transaction.SUCCESS);
+    } catch (Throwable ex) {
+      transaction.setStatus(ex);
+      logger.error("Rebuild cache failed", ex);
+    } finally {
+      transaction.complete();
+    }
+  }
+
+  private void loadNewAndUpdatedAccessKeys() {
+    boolean hasMore = true;
+    while (hasMore && !Thread.currentThread().isInterrupted()) {
+      //current batch is 500
+      List<AccessKey> accessKeys = accessKeyRepository
+          .findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(lastTimeScanned);
+      if (CollectionUtils.isEmpty(accessKeys)) {
+        break;
+      }
+
+      int scanned = accessKeys.size();
+      mergeAccessKeys(accessKeys);
+      logger.info("Loaded {} new/updated Accesskey from startTime {}", scanned, lastTimeScanned);
+
+      hasMore = scanned == 500;
+      lastTimeScanned = accessKeys.get(scanned - 1).getDataChangeLastModifiedTime();
+
+      if (hasMore) {
```

Comment:
It is better to add some comments here so that people will know why we need this logic here.

Reply:
Okay